pyomo/contrib/piecewise/transform/multiple_choice.py

```diff
@@ -0,0 +1,36 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.core.base import Transformation, TransformationFactory
+import pyomo.gdp.plugins.hull
+
+@TransformationFactory.register('contrib.piecewise.multiple_choice',
+                                doc="Convert piecewise-linear model to a GDP "
+                                "to 'Multiple Choice' MIP "
+                                "formulation.")
+class MultipleChoiceTransformation(Transformation):
+    """
+    Converts a model containing PiecewiseLinearFunctions to a an equivalent
+    MIP via the Multiple Choice method from [1]. Note that,
+    while this model probably resolves to the model described in [1] after
+    presolve, the Pyomo version is not as simplified.
```

Comment:
How is the model different? Can we simplify it here? Maybe some comments can help us understand how are these different.

Reply:
It's different because you don't actually have to disaggregate the variables that represent the value of the piecewise-linear function ('z 'in Vielma et al. and `substitute_var` here). To get the exact multiple choice formulation from Vilema et al., you just need to sum the set of equality constraints setting that var to the value of the linear function at (disaggregated) x. It turns out that's an equivalent formulation with fewer constraints and variables. We could simplify it here, but I'm going to wait to see if we need to: I'm assuming the model here is probably linear, so I expect there's nothing to do here that presolve can't do faster.